python/beeai_framework/emitter/emitter.py

```diff
@@ -194,31 +194,40 @@ async def emit(self, name: str, value: Any) -> None:
             raise EmitterError.ensure(e)
 
     async def _invoke(self, data: Any, event: EventMeta) -> None:
-        executions: list[Coroutine[Any, Any, Any] | Task[Any]] = []
-        for listener in self._listeners:
-            if not listener.match(event):
-                continue
-
-            if listener.options and listener.options.once:
-                self._listeners.remove(listener)
-
-            async def run(ln: Listener = listener) -> Any:
-                try:
-                    if inspect.iscoroutinefunction(ln.callback):
-                        return await ln.callback(data, event)
-                    else:
-                        return ln.callback(data, event)
-                except Exception as e:
-                    raise EmitterError.ensure(
-                        e, message="One of the provided Emitter callbacks has failed.", event=event
-                    )
+        tasks: list[Task[Any]] = []
+
+        async def run(ln: Listener) -> Any:
+            try:
+                ln_async = ensure_async(ln.callback)
+                return await ln_async(data, event)
+            except Exception as e:
+                raise EmitterError.ensure(
+                    e,
+                    message=f"One of the provided emitter callbacks has failed. Event: {event.path}",
+                    event=event,
+                )
 
-            if listener.options and listener.options.is_blocking:
-                executions.append(run())
-            else:
-                executions.append(asyncio.create_task(run()))
+        try:
+            for listener in self._listeners:
+                if not listener.match(event):
+                    continue
+
+                if listener.options and listener.options.once:
+                    self._listeners.remove(listener)
+
+                task = asyncio.create_task(run(listener))
+                tasks.append(task)
+
+                if listener.options and listener.options.is_blocking:
+                    _ = await task
+
+            await asyncio.gather(*tasks)
+        except:
+            for task in tasks:
+                task.cancel()
 
-        await asyncio.gather(*executions)
+            await asyncio.gather(*tasks, return_exceptions=True)
```

Comment:
consider using [TaskGroup](https://docs.python.org/3/library/asyncio-task.html#asyncio.TaskGroup), I think it handles some of these concerns better, such as auto cancellation:

> The first time any of the tasks belonging to the group fails with an exception other than [asyncio.CancelledError](https://docs.python.org/3/library/asyncio-exceptions.html#asyncio.CancelledError), the remaining tasks in the group are cancelled.